validation/superchain-version_test.go

```diff
@@ -0,0 +1,135 @@
+package validation
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	. "github.com/ethereum-optimism/superchain-registry/superchain"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+// TestContractVersions will check that
+//   - for each chain in OPChain
+//   - for each declared contract "Foo" : version entry in the corresponding superchain's semver.yaml
+//   - the chain has a contract FooProxy deployed at the same version
+//
+// Actual semvers are
+// read from the L1 chain RPC provider for the chain in question.
+func TestContractVersions(t *testing.T) {
+
+	isExcluded := map[uint64]bool{
+		10:           true,
+		291:          true,
+		420:          true,
+		424:          true,
+		888:          true,
+		957:          true,
+		997:          true,
+		8453:         true,
+		34443:        true,
+		58008:        true,
+		84531:        true,
+		84532:        true,
+		7777777:      true,
+		11155420:     true,
+		11763071:     true,
+		999999999:    true,
+		129831238013: true,
+	}
+
+	isSemverAcceptable := func(desired, actual string) bool {
+		return desired == actual
+	}
+
+	checkOPChainSatisfiesSemver := func(chain *ChainConfig) {
+		rpcEndpoint := Superchains[chain.Superchain].Config.L1.PublicRPC
+
+		if rpcEndpoint == "" {
+			t.Errorf("%s has MISSING RPC endpoint", chain.Superchain)
+			return
+		}
+
+		client, err := ethclient.Dial(rpcEndpoint)
+
+		if err != nil {
+			t.Errorf("could not dial rpc endpoint %s: %v", rpcEndpoint, err)
+			return
+		}
+
+		semverFields := reflect.VisibleFields(reflect.TypeOf(SuperchainSemver[chain.Superchain]))
+
+		for _, field := range semverFields {
+			desiredSemver := reflect.Indirect(reflect.ValueOf(SuperchainSemver[chain.Superchain])).FieldByName(field.Name).String()
```

Comment:
I find reflection to difficult to read and reason about. If we absolutely have to use it, let's hide the complexity behind utility getter methods on the structs containing the versions/addresses. Something like:

```go
// within superchain.go
func (c ContractVersions) VersionFor(name string) (string, error) {}

func (a AddressList) AddressFor(name string) (string, error) {}

// in the test

var requiredContracts = []string { "OptimismPortal", ... }
```

This way we avoid proliferating the use of `reflect` elsewhere in the codebase, and expose a cleaner more readable API to future developers.

Reply:
I went with your suggestion, only I didn't even use `reflect` in the implementations [9f78655](https://github.com/ethereum-optimism/superchain-registry/pull/62/commits/9f78655bdf19aa6a25542e48f663bb478d65a095).

I think there is maybe still a question about when to check for a missing address (at init time or at runtime). For now,  the new getters will return an error if the struct has a zero/default value for a field.